### DISCUSSION TITLE
How to handle abnormal/ectopic heart beats in peak detection?

### DISCUSSION THREAD

**Discussion by michielmeys** [neuropsychology/NeuroKit#1039] [Q&A]
I would like to learn, from the standpoint of R-peak detection for HRV analysis, how I should handle events in my ECG signals that correspond to abnormal/ectopic heart beats. In the image below is an example of the two main types of observations I am referring to. 

<img width="508" alt="3141_baseline_pre" src="https://github.com/user-attachments/assets/693ed14b-f64d-453d-803d-7c3b40943825">

The frequency of these occurrences in my data varies from only once in a span of 10 minutes to as much as once every three heart beats (in by far the most extreme case).

Depending on the direction and shape of these abnormal beats, a peak can sometimes be detected by the detection algorithm (see again image), however, do these still really "count" as beats and should any of these be used like that for the purpose of HRV analysis? Or would it be more correct to instead manually adjust such misplaced or missing beats based on the surrounding normal beats, in order to minimise their impact?

Apologies that this question is not really specific to the toolbox. If anybody knows a better place for me to get this answered, please let me know!

**Comment by DominiqueMakowski** [comment]
It is a good question. Ectopic beats and other detection issues are typically handled in a step called `ecg_fixpeaks()`, e.g.: 

https://github.com/neuropsychology/NeuroKit/blob/45c9ad90d863ebf4e9d043b975a10d9f8fdeb06b/neurokit2/signal/signal_fixpeaks.py#L331-L333

which is used by `ecg_peaks()`:

https://github.com/neuropsychology/NeuroKit/blob/45c9ad90d863ebf4e9d043b975a10d9f8fdeb06b/neurokit2/ecg/ecg_peaks.py#L258-L264

(which output is then used by the HRV functions).

It is somewhat imperfect, and a manual data quality check is recommended

**Reply by michielmeys** [reply]
OK, just a pity that some clearly correctly detected normal beats will also be affected by this sometimes, but I suppose that is the tradeoff then. I will perhaps try to use `signal_fixpeaks()` afterwards outside of `ecg_peaks()` so I can consider the ones flagged as ectopic only. Thank you for your reply
